### ISSUE TITLE
focusAtom type error

### ISSUE THREAD

**Issue by ViktorQvarfordt** [jotaijs/jotai-optics#6]
The [example code from the docs](https://jotai.org/docs/integrations/optics#focusatom)

```ts
const baseAtom = atom({ a: 5 }) // PrimitiveAtom<{a: number}>
const derivedAtom = focusAtom(baseAtom, (optic) => optic.prop('a')) // PrimitiveAtom<number>
```

gives this typescript error:

```
No overload matches this call.
  The last overload gave the following error.
    Argument of type 'PrimitiveAtom<{ a: number; }> & WithInitialValue<{ a: number; }>' is not assignable to parameter of type 'WritableAtom<{ a: number; } | undefined, [{ a: number; }], void>' with 'exactOptionalPropertyTypes: true'. Consider adding 'undefined' to the types of the target's properties.
      Types of property 'onMount' are incompatible.
        Type 'OnMount<SetStateAction<{ a: number; }>, void>' is not assignable to type 'OnMount<[{ a: number; }], void>'.
          Types of parameters 'setAtom' and 'setAtom' are incompatible.
            Type 'S' is not assignable to type '(update: SetStateAction<{ a: number; }>) => void'.
              Type 'SetAtom<[{ a: number; }], void>' is not assignable to type '(update: SetStateAction<{ a: number; }>) => void'.
                Types of parameters 'args_0' and 'update' are incompatible.
                  Type 'SetStateAction<{ a: number; }>' is not assignable to type '{ a: number; }'.
                    Type '(prev: { a: number; }) => { a: number; }' is not assignable to type '{ a: number; }'.ts(2769)
focusAtom.d.ts(15, 25): The last overload is declared here.
```

Getting a similar error message also after disabling `exactOptionalPropertyTypes` as the error message suggests:

```
No overload matches this call.
  The last overload gave the following error.
    Argument of type 'PrimitiveAtom<{ a: number; }> & WithInitialValue<{ a: number; }>' is not assignable to parameter of type 'WritableAtom<{ a: number; } | undefined, [{ a: number; }], void>'.
      Types of property 'onMount' are incompatible.
        Type 'OnMount<SetStateAction<{ a: number; }>, void> | undefined' is not assignable to type 'OnMount<[{ a: number; }], void> | undefined'.
          Type 'OnMount<SetStateAction<{ a: number; }>, void>' is not assignable to type 'OnMount<[{ a: number; }], void>'.
            Types of parameters 'setAtom' and 'setAtom' are incompatible.
              Type 'S' is not assignable to type '(update: SetStateAction<{ a: number; }>) => void'.
                Type 'SetAtom<[{ a: number; }], void>' is not assignable to type '(update: SetStateAction<{ a: number; }>) => void'.
                  Types of parameters 'args_0' and 'update' are incompatible.
                    Type 'SetStateAction<{ a: number; }>' is not assignable to type '{ a: number; }'.
                      Type '(prev: { a: number; }) => { a: number; }' is not assignable to type '{ a: number; }'.ts(2769)
focusAtom.d.ts(15, 25): The last overload is declared here.
```

Screenshot for reference:
<img width="1201" alt="image" src="https://user-images.githubusercontent.com/344809/216185673-e3c14ed7-394b-4e73-a05f-995a1f80c98b.png">

Using these versions
```
    "jotai": "^2.0.0",
    "jotai-optics": "^0.3.0",
    "optics-ts": "^2.4.0",
```

**Comment by ViktorQvarfordt** [comment]
Localized the issue to `"moduleResolution": "nodenext"` in tsconfig.

[Here](https://stackblitz.com/edit/typescript-pyt3aa?file=index.ts) is a type of minimal reproducible example. It doesn't show the same verbose error message as in my post above, but it does show that the type does not get inferred correctly.

**Comment by dai-shi** [comment]
Works fine here: https://tsplay.dev/WJ7L6w
<img width="634" alt="image" src="https://user-images.githubusercontent.com/490574/216193496-80503237-762b-43cf-b331-775d3747425d.png">

So, it's a bundler/bundling issue??

**Comment by ViktorQvarfordt** [comment]
It loks like that site doesn’t allow to change the tsconfig.

I wouldn’t say it’s a bundler issue but rather an issue that occurs under certain typescript settings.

**Comment by dai-shi** [comment]
so, how can we fix it on the library end?

**Comment by arjunvegda** [comment]
I haven't looked too deeply into this, but could the root cause be `optics-ts`?

When module resolution is set to `node` and `optics-ts` resolves the types correctly. See - https://tsplay.dev/wgAP0m 

When module resolution is set to `nodenext`, `optics-ts` doesn't resolve the types correctly. See - https://tsplay.dev/WYRzdw . It appears to be resolving to the `cjs` module but can't find types for it.

<img width="776" alt="image" src="https://user-images.githubusercontent.com/14841132/216225933-c497cb05-9af6-4195-9a61-8c43b356298b.png">

**Comment by cristatus** [comment]
I was facing the same issue with `moduleResolution: "bundler"`. I tried patching the `optics-ts` to export types declaration file from the `exports` section. It fixed the `unknown` type issue.

```json
 "exports": {
     ".": {
       "types": "./index.d.ts",
       "import": "./dist/mjs/index.js",
       "require": "./dist/cjs/index.js"
     },
```

 but than I saw `Type instantiation is excessively deep and possibly infinite.` error with `focusAtom` function.

This issue is fixed by patching the `dist/src/focusAtom.d.ts` by replacing `O.OpticFor<S>` with `O.OpticFor_<S>`.

Will try to submit PR to respective repos.

**Comment by Aleksion** [comment]
@cristatus did you submit a PR for it? focusAtom is basically killing my vscode editor because of this 😂

**Comment by cristatus** [comment]
Just created here #10 and akheron/optics-ts#318

**Comment by arjunvegda** [comment]
@Aleksion - Would you like to test it on your end with the canary version? [Here are the instructions](https://github.com/jotaijs/jotai-optics/pull/10#issuecomment-1627243966) to install @cristatus's solution

**Comment by Aleksion** [comment]
I did - and it solved it. (I forked and tried it manually but didn't get to push it).

So all good from here 🎉

**Comment by arjunvegda** [comment]
I just checked that [this PR (from @cristatus on optics-ts](https://github.com/akheron/optics-ts/pull/318) repo is merged. I think we should wait for it to be released and see if that fixes the issue 🤔

**Comment by arjunvegda** [comment]
Are we good to close this issue? Did https://github.com/akheron/optics-ts/pull/318 fix it?

**Comment by arjunvegda** [comment]
The fix is released in v0.3.1 🚀
